subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -102,99 +110,167 @@ def check_payments(self, payments: Iterable[SubscriptionPayment]):
+
+        # This gets us the very first transaction that happened for this person.
         subscription_payment = SubscriptionPayment.objects.get(
             provider_codename=self.codename,
-            provider_transaction_id=payload.transaction_info.original_transaction_id,
+            provider_transaction_id=transaction_info.original_transaction_id,
         )
 
-        # Currently, we don't support changing of the product ID. Assert here will let us know if anyone did that.
-        # In case the field is not available in metadata, the product ID error will still be raised.
+        # Information about this should've been received earlier via change, but downgrades are to happen
+        # only at the start of new cycle. So we handle them here.
         current_product_id = subscription_payment.plan.metadata.get(self.codename)
-        if current_product_id != payload.transaction_info.product_id:
-            raise ProductIdChangedError(current_product_id, payload.transaction_info.product_id)
+        if current_product_id != transaction_info.product_id:
+            subscription_payment.plan = self._get_plan_for_product_id(transaction_info.product_id)
+            # Create a new subscription.
+            subscription_payment.subscription = None
 
         subscription_payment.pk = None
         # Updating relevant fields.
-        subscription_payment.provider_transaction_id = payload.transaction_info.transaction_id
-        subscription_payment.subscription_start = payload.transaction_info.purchase_date
-        subscription_payment.subscription_end = payload.transaction_info.expires_date
+        subscription_payment.provider_transaction_id = transaction_info.transaction_id
+        subscription_payment.subscription_start = transaction_info.purchase_date
+        subscription_payment.subscription_end = transaction_info.expires_date
         subscription_payment.save()
 
         return Response(status=HTTP_200_OK)
+
+    def _handle_subscription_change(self, user: AbstractBaseUser, notification: AppStoreNotification) -> Response:
+        assert notification.subtype in {
+            AppStoreNotificationTypeV2Subtype.UPGRADE,
+            AppStoreNotificationTypeV2Subtype.DOWNGRADE,
+        }, f'Unsupported notification subtype received for subscription change: {notification.subtype}.'
+
+        if notification.subtype == AppStoreNotificationTypeV2Subtype.DOWNGRADE:
+            logger.info('User %s requested a downgrade to %s. This will happen during the next renewal.', )
+            return Response(status=HTTP_200_OK)
+
+        # We handle this in two steps:
+        # 1. We shorten the current subscription and subscription payment to current start of the period.
+        # 2. We add a new subscription with a new plan.
+
+        # Find the latest apple payment by this user.
+        latest_payment = SubscriptionPayment.objects.filter(
+            provider_codename=self.codename,
+            user=user,
+        ).latest()
```

Comment:
Currently payments are sorted for `latest()` by `created` field which is stupid - the field should be `end`. I will fix later.
Here better use `order_by(...).last()` or something.